### DISCUSSION TITLE
ConverterScaler not changing the volume correctly

### DISCUSSION THREAD

**Discussion by GustavAbrahamsson** [pschatzmann/arduino-audio-tools#157] [Q&A] [answered]
Hi! Amazing package this is, I'm quite new to this so I've come across some problems. One of them is the fact that ConverterScaler doesn't control the volume smoothly from 0.0 to 1.0. In Volume Control in the bottom-most example, we have this line:

`ConverterScaler<int16_t> volume(1.0, 0, 32767);            // volume control`

When using this i my code, I could see that .setFactor() only changed the factor to 0.0 or 1.0, even if I just manually wrote in 0.5. I then switched the <int16_t> to <float> and it did change the volume as .setFactor() got arguments ∈ (0.0, 1.0) (and .volume() returned the same values!), but it was almost completely random. Sweeping from 0.0 to 1.0 made the speaker go nuts, going up and down in volume.

I'm just trying to control the volume of a generated square wave. main.cpp is littered with old and out-commented code, so here are only the relevant parts:

```
[...]
#define MAX_VOLUME 150
typedef int16_t sound_t;

SquareWaveGenerator<sound_t> squareWaveE(MAX_VOLUME);
GeneratedSoundStream<sound_t> soundE(squareWaveE);
I2SStream i2s;
StreamCopy copier2(i2s, soundE);

float volFactor = 1.0; // Controlled by a potentiometer
ConverterScaler<float> volume(volFactor, 0, MAX_VOLUME); // Note: it was <int16_t> in the example

[...]

void setup(void) {
  I2SConfig config = i2s.defaultConfig(TX_MODE);
  config.sample_rate = sample_rate; 
  config.channels = channels;
  config.bits_per_sample = 16;
  i2s.begin(config);
  
  squareWaveE.begin(channels, sample_rate, 100);
}
 
void loop() {
  [volFactor gets assigned a value between 0.0 and 1.0 by the pot]

  if(abs(volume.factor() - volFactor) > 0.05) volume.setFactor(volFactor); // adjust the volume with some deadzone to stabilize

  copier2.copy(volume);
}
```

I'm running an ESP32 lolin32 with a MAX98357 amplifier over I2S along with a basic speaker. I also tried implementing effects and using a boost pedal, but I couldn't even get a sound even though .volume() returned values >0.

Grateful for all the help I can get.

**Comment by pschatzmann** [comment]
I suggest that you read the documentation about [volume](https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control).

In your code the ```ConverterScaler<float>``` is wrong because the indicated data type does not match with the audio data.

**Reply by GustavAbrahamsson** [reply]
It was the bottom-most example in Volume Control that I based my code upon. I'm trying to understand it. It gave either 1.0 or 0.0 volume, nothing inbetween. Changing it from `ConverterScaler<int16_t>` to `ConverterScaler<float>` made it possible to get volumes _between_ 0.0 and 1.0, but it wasn't "in order". I.e `volume.setFactor(0.1)` might result in the speaker being louder than `volume.setFactor(0.2)` (verified with volume.factor() ).

**Comment by pschatzmann** [comment]
As I have indicated ```ConverterScaler<float>``` is plain wrong. Int16_t values are from -32767 to 32767 so MAX_VOLUME should be  32767. So your range is also much too small!
Please also note that the values are applied just as simple multiplication (which gives as described a bad audible scaling)!   

I suggest that you just stick to the example and it will work...

**Reply by GustavAbrahamsson** [reply]
Oh, it's not advisable to `SquareWaveGenerator<sound_t> squareWaveE(MAX_VOLUME)` where MAX_VOLUME < 32767? The speaker was reeally loud and I didn't want to blow up my ears so I just chose a different parameter when creating the generators. I assume that the only way to make it more quiet is with volume control, and not a combination?

Thanks for the help, I'll get back to you when I get to try it out!

**Comment by pschatzmann** [comment]
Yes it can be lower - but you want to give much more range to control. 
That's the reason why I would recommend to use VolumeStream: If you look at the curve it does not get loud so fast...

**Reply by GustavAbrahamsson** [reply]
Ok! In your example you had `EncodedAudioStream dec(&volume, new MP3DecoderHelix()); // Decoding stream`, but I only have a square wave I want to output on I2S. What more has to be done except adding `VolumeStream volume(i2s);`? It has to be linked to the output in some way, right?

**Comment by pschatzmann** [comment, marked as answer]
The following should work
```
SquareWaveGenerator<sound_t> squareWaveE(MAX_VOLUME);
GeneratedSoundStream<sound_t> soundE(squareWaveE);
VolumeStream volume(soundE);
I2SStream i2s;
StreamCopy copier2(i2s, volume);
```

**Reply by GustavAbrahamsson** [reply]
Alright! I'll get back to you when I've tried it. Thanks so far!

**Reply by GustavAbrahamsson** [reply]
Thanks, it worked!

**Reply by palomena** [reply]
Hi Gustav,  

sorry for re-activating this thread, but I noticed you have a similar setup to mine and had issues with the volume control. Maybe you can help me figure out whether I should open an issue or if my problems are amp/speaker related.  
I too use the MAX98356a I2S amplifier with an esp32, this great library, a 3 W, 4 ohm speaker and volume control using a potentiometer. Everything works as expected, however when I change the volume the slightest bit I get a noticable "pop" out of the speaker. It is most noticable when setting the volume to zero and back to "normal". Very annoying as i need to create a beeping sound.
Implementing a gradual volume change that approaches the potentiometer setting by very small increments fixes the problem for both volume and frequency, however this is too slow for a beep. Even a 0.05% change in volume results in a  noticable pop. I use a SineWaveGenerator fed into a VolumeStream on the software side.   
Now my question: Did you experience similar issues or can you change the volume without any pop occuring? Just looking for your experience regarding changing the volume. Thanks

**Comment by pschatzmann** [comment]
Did you try to debounce your input ?

**Reply by palomena** [reply]
Hi and thank you for your response!  
I have a capacitor in line with the potentiometer and use a simple moving average filter to smooth out the analog results.  
The problem is most definitely not a noisy volume control though. I have verified the volume i am setting and it's smooth and in small increments. For the beeping effect (switching volume between 0.0f and 1.0f or 0.5f) I do not use an external volume control at all. It is all done in software in a predictable manner:
```c
	static bool enabled = false;
	static unsigned long lastTime = 0;
	unsigned long ctime = millis();
	if(ctime > lastTime + 1000) {
		enabled = !enabled;
		lastTime = ctime;
	}

	if(enabled) {
		this->volume->setVolume(1.0f);
	} else {
		// mute
		this->volume->setVolume(0.0f);
	}
```  
Today I have tried swapping the SineWaveGenerator for a SquareWaveGenerator and the noise / pops when changing the volume seem to disappear (its a bit noisier than a sin wave so i might be missing some faint pops, but overall its a lot better). I therefore assume it is a problem with the SineWaveGenerator, but I am not sure and it sounds unlikely, since the VolumeStream has nothing to do with the generator itself (it just copies and alters the amplitude?) and therefore something like an unintended phase shift or similar problems seem unlikely. Initially I have assumed the angle of attack when changing volume is too steep, but that would also affect the SquareWave I guess.  
I have also tried the player class with a WAV file but I (or the library) seem to have issues with the SD interface which is i a problem of itself (not relevant here, might be another issue if i cannot get that fixed, though more likely to be an SD card issue, since they are not known for their interoperability on the SPI bus... :/).  
To sum it up once more:  
I cannot get the SineWaveGenerator to change volume without popping. Especially when muting and unmuting it in certain time intervals.  
Since I am not sure whether it is related to the MAX98357 I2S amplifier, I decided to ask for experiences of other users with that setup. A quick internet search turned up some results indicating that it could be amplifier related ("MAX98357 popping" - mostly raspberry pi and volumio users).

**Comment by pschatzmann** [comment]
If I understand your logic correctly you switch from the volume 1.0 to 0 from one moment to the other. If the signal is at that point not zero (which is very likely), you generate automatically a pop! 
- The way to avoid this is to switch off when the sine wave is crossing zero (which might be quite difficult)
- To decrement the volume over some time

**Reply by palomena** [reply]
I tried the 2nd option - this is the gradual change i was talking about. While this decreases the volume of these pops, it increases the amount of pops significantly. For every change, however small, there is a faint pop. And any change under *0.01f* in Volume takes way too much time, especially when dealing with high amplitudes.  
Some pseudocode:  
```c
each second {
      enabled = !enabled;
}

// volume is the volume setting (what volume should be)
// actualVolume is the current volume (what volume is)
if(enabled) volume = 0.5f;
else           volume = 0.0f;

if(volume != actualVolume) {
		// flcmp compares floats and returns -1 if less, 1 if greater, 0 if equal within certain epsilon distance
		int c = flcmp(volume, actualVolume);
		if(c > 0) {
			actualVolume += 0.025f; // adjust volume by a small delta to approach desired volume setting
		} else if(c < 0) {
			actualVolume -= 0.025f; // adjust by a small delta
		}
		this->volumeStream->setVolume(actualVolume);
		Serial.print("Volume: "); Serial.print(volume);
		Serial.print(" , "); Serial.println(actualVolume);
	}
```

And I do get some nice output, indicating a successful approach. Still "popping" though. On each volume adjustment. Sounds like a tiny bee is trapped inside the speaker. Maybe there is, who knows... :)
```
Volume: 0.00 , 0.00
Volume: 0.00 , 0.00
Volume: 0.00 , 0.00
Volume: 0.00 , 0.00
Volume: 0.30 , 0.03
Volume: 0.30 , 0.05
Volume: 0.30 , 0.08
Volume: 0.30 , 0.10
Volume: 0.30 , 0.13
Volume: 0.30 , 0.15
Volume: 0.30 , 0.18
Volume: 0.30 , 0.20
Volume: 0.30 , 0.23
Volume: 0.30 , 0.25
Volume: 0.30 , 0.28
Volume: 0.30 , 0.30
Volume: 0.30 , 0.30
```  

Adding a capacitor isn't much help too. It slightly improves the popping/buzzing, but degrades sound quality of course.   
Any idea how I could try the first option, i.e. detecting an angle of 0° in the sine wave? Thank you for your valuable input so far, much appreciated!

**Comment by pschatzmann** [comment]
You can decrease the buffer size e.g. to 4 bytes when you define the copier. 
This should resolve your time issue...

**Reply by palomena** [reply]
Strangely enough that is still not making things (much) better.    
I have recorded a small sample with these settings:  
delta Volume Adjustment: 0.00025f  
max Amplitude: 8000  
buffersize: 4 Bytes (tried 2^n power between 4 and 256, all the same)   
[Link to sndup.net](http://www.sndup.net/852w)  
I suggest headphones or increasing the volume a little, otherwise it may be hard to hear. In person it is easier to notice and harder to not notice. The speaker is floating in the air, so no surface noise.   
[Speaker model](https://www.amazon.de/Metallgeh%C3%A4use-Runder-Magnet-Lautsprecher-St%C3%BCck/dp/B00NQ0LHNA/ref=sr_1_4?__mk_de_DE=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=1KQK18QT6SBPD&keywords=3+watt+4+ohm+speaker&qid=1649778214&sprefix=3watt+4+ohm+speake%2Caps%2C135&sr=8-4)  
I understand if you are not able to provide further input, as this might be unrelated to your library. Did you ever try outputting an i2s sine wave and changing the volume of that or muting it? The last straw would be detecting a phase of 0° now, i guess...

**Comment by pschatzmann** [comment]
My usual use case (with the the AudioKit) is to use the volume up and volume down buttons which change the volume in steps.
I suggest that you copy and adjust the ConverterScaler, where you could try to find filter out the final step with a big gap...

You could also use the SineFromTable and instead of setting the volume to 0 set the setFrequency to 0 to silence it.
I just added and committed the setAmplitude() method to SineFromTable. SineFromTable is changing settings at 360 degrees (which is the zero point)

**Reply by palomena** [reply]
Yup, that new SineFromTable function combined with a gradual change in volume (this is still required since any sudden change in volume still results in a sudden mechanical reaction of the speaker thus producing a pop) fixes my problem. The zero-crossing-sentient setAmplitude function reduces our required volume delta steps to avoid pops by several orders of magnitude (and without it it wouldn't even work with miniscule step sizes). Verified with the MAX98357 amplifier, the CJMCU-5102 and the UDA1334, all working fine with sine waves now.  
Thank you for the quick solution, I really did not think about the sine wave basically convoluting with a delta function for a short amount of time when changing the gain at a non-zero phase position. Probably would have bugged me the whole week until I'd finally given up and head to the lab to throw the Oscilloscope at it... Might be worth an entry in the Wiki - maybe I'll contribute something... I don't think I'll be the last one to stumble upon this. I am probably not the only one using your library to avoid most of those darn Signal Processing details.  
Again big thank you for the help and my compliments with reference to the library - I like it a lot!

**Comment by pschatzmann** [comment]
I tried to add this logic into the setAmplitude() method. The step size can be defined with setMaxAmplitudeStep(). Currently I am using 50 as default value.

**Reply by fobi55** [reply]
Hello pschatzmann, thanks for the library!

There is a small error in ConverterScaler class, the setFactor type is not right.

`/// Defines the factor (volume)
//void setFactor(T factor) { this->factor_value = factor; }
void setFactor(float factor) { this->factor_value = factor; }`

Apart of this, it is working as expected:

`void loop() {

  if (counter > 1000) {
    counter = 0;
    volFactor = volFactor + 0.1;
    if(volFactor > 1){volFactor = 0;}
    vol.setFactor(volFactor);
    Serial.println(vol.factor());

  }

  copier.copy(vol);
}`

Thanks.

**Comment by pschatzmann** [comment]
You are right: setFactor should be float.
I committed a correction, however I suggest to prefer to use the VolumeStream class instead to manage the volume

**Reply by fobi55** [reply]
Yes, I got this, thank you.
I am interested in channel manipulation for SDR, lets say FM detection: 
m(t)=scale*(I(t)*delay{Q(t)}-Q(t)*delay{I(t)}), where I and Q are L and R channels. 
And because the ConverterScaler class has offset:
`void setOffset(T offset) { this->offset_value = offset; }`, I decided to try it as delay.
I think your library is powerful enough to do streams math manipulation without another DSP library or new subclasses.
Thank you again.
This is outside the scope of the discussion, sorry, I'm making excuses.
B

**Reply by fobi55** [reply]
Oh, the `T offset` is just a DC, my mistake.
